node/src/tasks/msp_upload_file.rs

```diff
@@ -0,0 +1,854 @@
+};
+use shc_file_manager::traits::{FileStorageWriteError, FileStorageWriteOutcome};
+use shc_file_transfer_service::{
+    commands::FileTransferServiceInterface, events::RemoteUploadRequest,
+};
+use shc_forest_manager::traits::ForestStorage;
+use storage_hub_runtime::StorageDataUnit;
+
+const LOG_TARGET: &str = "msp-upload-file-task";
+
+const MAX_CONFIRM_STORING_REQUEST_TRY_COUNT: u32 = 3;
+
+/// MSP Upload File Task: Handles the whole flow of a file being uploaded to a MSP, from
+/// the MSP's perspective.
+///
+/// The flow is split into three parts, which are represented here as 3 handlers for 3
+/// different events:
+/// - [`RemoteUploadRequest`] event: The second part of the flow. It is triggered by a
+///   user sending a chunk of the file to the MSP. It checks the proof for the chunk
+///   and if it is valid, stores it, until the whole file is stored. Finally the MSP will
+///   construct a forest proof of non-inclusion and a file proof and send a accept transaction
+///   to StorageHub which will automatically apply the delta to update the bucket root.
+///   If successful, the MSP will also update their local forest to include the new file.
+pub struct MspUploadFileTask<FL, FSH>
+where
+    FL: FileStorageT,
+    FSH: MspForestStorageHandlerT,
+{
+    storage_hub_handler: StorageHubHandler<FL, FSH>,
+    file_key_cleanup: Option<H256>,
+}
+
+impl<FL, FSH> Clone for MspUploadFileTask<FL, FSH>
+where
+    FL: FileStorageT,
+    FSH: MspForestStorageHandlerT,
+{
+    fn clone(&self) -> MspUploadFileTask<FL, FSH> {
+        Self {
+            storage_hub_handler: self.storage_hub_handler.clone(),
+            file_key_cleanup: self.file_key_cleanup,
+        }
+    }
+}
+
+impl<FL, FSH> MspUploadFileTask<FL, FSH>
+where
+    FL: FileStorageT,
+    FSH: MspForestStorageHandlerT,
+{
+    pub fn new(storage_hub_handler: StorageHubHandler<FL, FSH>) -> Self {
+        Self {
+            storage_hub_handler,
+            file_key_cleanup: None,
+        }
+    }
+}
+
+/// Handles the [`NewStorageRequest`] event.
+///
+/// This event is triggered by an on-chain event of a user submitting a storage request to StorageHub.
+///
+/// This task will:
+/// - Check if the MSP has enough storage capacity to store the file and increase it if necessary (up to a maximum).
+/// - Register the user and file key in the registry of the File Transfer Service, which handles incoming p2p
+/// upload requests.
+impl<FL, FSH> EventHandler<NewStorageRequest> for MspUploadFileTask<FL, FSH>
+where
+    FL: FileStorageT,
+    FSH: MspForestStorageHandlerT,
+{
+    async fn handle_event(&mut self, event: NewStorageRequest) -> anyhow::Result<()> {
+        info!(
+            target: LOG_TARGET,
+            "Registering user peer for file_key {:?}, location {:?}, fingerprint {:?}",
+            event.file_key,
+            event.location,
+            event.fingerprint
+        );
+
+        let result = self.handle_new_storage_request_event(event).await;
```

Comment:
Why did you externalize this logic into a function if it's only used once?

Reply:
I copied the same pattern used in the BSP task just to keep it consistent.